File: app/src/main/java/com/example/david/lists/view/itemlist/buldlogic/ItemListViewModule.java
package com.example.david.lists.view.itemlist.buldlogic;

import android.app.Application;

import androidx.recyclerview.widget.ItemTouchHelper;

import com.chauthai.swipereveallayout.ViewBinderHelper;
import com.example.david.lists.common.buildlogic.ViewScope;
import com.example.david.lists.data.repository.IRepositoryContract;
import com.example.david.lists.util.ISchedulerProviderContract;
import com.example.david.lists.view.itemlist.IItemViewContract;
import com.example.david.lists.view.itemlist.ItemAdapter;
import com.example.david.lists.view.itemlist.ItemListLogic;
import com.example.david.lists.view.itemlist.ItemListViewModel;

import dagger.Module;
import dagger.Provides;
import io.reactivex.disposables.CompositeDisposable;

@Module
final class ItemListViewModule {
    @ViewScope
    @Provides
    IItemViewContract.Logic logic(IItemViewContract.View view,
                                  IItemViewContract.ViewModel viewModel,
                                  IRepositoryContract.Repository repo,
                                  ISchedulerProviderContract schedulerProvider,
                                  CompositeDisposable disposable) {
        return new ItemListLogic(view, viewModel, repo, schedulerProvider, disposable);
    }

    @ViewScope
    @Provides
    IItemViewContract.ViewModel viewModel(Application application, String userListId) {
        return new ItemListViewModel(application, userListId);
    }

    @ViewScope
    @Provides
    IItemViewContract.Adapter adapter(IItemViewContract.Logic logic,
                                      ViewBinderHelper viewBinderHelper,
                                      ItemTouchHelper itemTouchHelper) {
        return new ItemAdapter(logic, viewBinderHelper, itemTouchHelper);
    }
}

File: app/src/main/java/com/example/david/lists/view/itemlist/buldlogic/ItemListViewModule.kt
package com.example.david.lists.view.itemlist.buldlogic

import android.app.Application
import androidx.recyclerview.widget.ItemTouchHelper
import com.chauthai.swipereveallayout.ViewBinderHelper
import com.example.david.lists.common.buildlogic.ViewScope
import com.example.david.lists.util.ISchedulerProviderContract
import com.example.david.lists.view.itemlist.IItemViewContract
import com.example.david.lists.view.itemlist.ItemAdapter
import com.example.david.lists.view.itemlist.ItemListLogic
import com.example.david.lists.view.itemlist.ItemListViewModel
import com.example.domain.repository.IRepositoryContract
import dagger.Module
import dagger.Provides
import io.reactivex.disposables.CompositeDisposable

@Module
internal class ItemListViewModule {
    @ViewScope
    @Provides
    fun logic(view: IItemViewContract.View,
              viewModel: IItemViewContract.ViewModel,
              repo: IRepositoryContract.Repository,
              schedulerProvider: ISchedulerProviderContract,
              disposable: CompositeDisposable): IItemViewContract.Logic {
        return ItemListLogic(view, viewModel, repo, schedulerProvider, disposable)
    }

    @ViewScope
    @Provides
    fun viewModel(application: Application, userListId: String): IItemViewContract.ViewModel {
        return ItemListViewModel(application, userListId)
    }

    @ViewScope
    @Provides
    fun adapter(logic: IItemViewContract.Logic,
                viewBinderHelper: ViewBinderHelper,
                itemTouchHelper: ItemTouchHelper): IItemViewContract.Adapter {
        return ItemAdapter(logic, viewBinderHelper, itemTouchHelper)
    }
}
